sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala

```diff
@@ -497,6 +498,11 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
       StructType(schema.filterNot(_.name == parsedOptions.columnNameOfCorruptRecord))
 
     val linesWithoutHeader: RDD[String] = maybeFirstLine.map { firstLine =>
+      if (parsedOptions.enforceSchema == false) {
+        CSVDataSource.checkHeader(firstLine, new CsvParser(parsedOptions.asParserSettings),
```

Comment:
Does this logic ever validate each file's header at all?

Reply:
It should check CSV header each time when they are removed. Do you think I missed some cases?